<?xml version="1.0"?>

<launch>
  <arg name="robot_name" default="$(optenv TIANRACER_NAME /)" doc="robot name [tianracer_No1, tianracer_No2, tianracer_No3, ...]"/>
  <arg name="serial_port" default="$(optenv TIANRACER_BASE_PORT /dev/ttyUSB1)" />
  <arg name="serial_baudrate" default="$(optenv TIANRACER_BASE_BAUDRATE 115200)" />

  <arg name="base_frame" default="base_link"/>
  <arg name="odom_frame" default="odom"/>
  <arg name="imu_frame" default="imu_link"/>

  <arg name="map_frame" default="map"/>
  <arg name="base_link_frame" default="base_footprint"/>
  <arg name="world_frame" default="odom"/>
  <arg name="odom0" default="odom"/>
  <arg name="imu0" default="imu"/>

  <!-- drivers for tianracer and tianbot are combined -->
  <group ns="$(arg robot_name)">
    <node pkg="tianbot_core" type="tianbot_core" name="tianracer" output="screen">
      <param name="serial_port" type="string" value="$(arg serial_port)"/>
      <param name="serial_baudrate" type="int" value="$(arg serial_baudrate)"/>
      <!-- there is no device type check for mini board -->
      <param name="type" type="string" value="ackermann" />
      <param if="$(eval serial_baudrate==115200)" name="type_verify" type="bool" value="false" />
      <param name="publish_tf" type="bool" value="false" />

      <!--if we have a group, we need to change the frames accordingly -->
      <param unless="$(eval robot_name=='/')" name="base_frame" value="$(arg robot_name)/$(arg base_frame)"/>
      <param unless="$(eval robot_name=='/')" name="odom_frame" value="$(arg robot_name)/$(arg odom_frame)"/>
      <param unless="$(eval robot_name=='/')" name="imu_frame" value="$(arg robot_name)/$(arg imu_frame)"/>

      <!-- tianbot_core currently using a private node -->
      <remap unless="$(eval robot_name=='/')" from="tianracer/ackermann_cmd" to="ackermann_cmd"/>
      <remap unless="$(eval robot_name=='/')" from="tianracer/debug_cmd" to="debug_cmd" />
      <remap unless="$(eval robot_name=='/')" from="tianracer/debug_result" to="debug_result" />
      <remap unless="$(eval robot_name=='/')" from="tianracer/imu" to="imu" />
      <remap unless="$(eval robot_name=='/')" from="tianracer/odom" to="odom" />
      <remap unless="$(eval robot_name=='/')" from="tianracer/uwb" to="uwb" />
    </node>

    <!-- Robot_Localization for odometry ekf fusion-->
    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_se" clear_params="true">
        <rosparam command="load" file="$(find tianracer_core)/param/tianbot_ekf_params.yaml" />
        <!-- <param unless="$(eval robot_name=='/')" name="map_frame" value="$(arg robot_name)/$(arg map_frame)"/>
        <param unless="$(eval robot_name=='/')" name="odom_frame" value="$(arg robot_name)/$(arg odom_frame)"/>
        <param unless="$(eval robot_name=='/')" name="base_link_frame" value="$(arg robot_name)/$(arg base_link_frame)"/>
        <param unless="$(eval robot_name=='/')" name="world_frame" value="$(arg robot_name)/$(arg world_frame)"/> -->
        <!-- Guarantee consistency of the topics with tianracer with namespaces  -->
        <param unless="$(eval robot_name=='/')" name="odom0" value="$(arg odom0)"/>
        <param unless="$(eval robot_name=='/')" name="imu0" value="$(arg imu0)"/>
    </node>
  </group>

</launch>
